library/core/src/str/validations/utf8_dfa.rs

```diff
@@ -0,0 +1,324 @@
+/// returning `Ok(())` in that case, or, if it is invalid, `Err(err)`.
+#[inline]
+#[rustc_const_unstable(feature = "str_internals", issue = "none")]
+pub(crate) const fn run_utf8_validation(inp: &[u8]) -> Result<(), Utf8Error> {
+    const fn validate_utf8_const(inp: &[u8]) -> Result<(), Utf8Error> {
+        validate_carefully(inp, 0)
+    }
+
+    #[inline]
+    fn validate_utf8_rt(inp: &[u8]) -> Result<(), Utf8Error> {
+        match inp.len() {
+            0 => Ok(()),
+            1..=CHUNK_LEN => {
+                if is_ascii_small(inp) {
+                    Ok(())
+                } else {
+                    validate_carefully(inp, 0)
+                }
+            }
+            _ => validate_utf8_impl(inp),
+        }
+    }
+
+    // SAFETY: These are equivalent aside from optimizations.
+    unsafe { const_eval_select((inp,), validate_utf8_const, validate_utf8_rt) }
+}
+
+/// Validation code which isn't particularly optimized, but can produce an
+/// accurate `Utf8Error` (and works in `const`), but
+#[inline]
+const fn validate_carefully(input: &[u8], mut pos: usize) -> Result<(), Utf8Error> {
+    let mut state = END;
+    let mut valid_up_to = pos;
+    while pos < input.len() {
+        state = dfa_step(state, input[pos]);
+        match state {
+            END => valid_up_to = pos + 1,
+            ERR => {
+                debug_assert!(pos >= valid_up_to && pos - valid_up_to <= 3);
+                let error_len =
+                    Some(if pos == valid_up_to { 1 } else { (pos - valid_up_to) as u8 });
+                return Err(Utf8Error { valid_up_to, error_len });
+            }
+            // Keep going
+            _ => {}
+        }
+        pos += 1;
+    }
+    if state != END { Err(Utf8Error { valid_up_to, error_len: None }) } else { Ok(()) }
+}
+
+#[inline]
+fn validate_utf8_impl(inp: &[u8]) -> Result<(), Utf8Error> {
+    let mut state = END;
+    let mut pos = 0;
+    let (chunks, tail) = inp.as_chunks::<CHUNK_LEN>();
+    if !chunks.is_empty() {
+        let mut last_state = state;
+        let mut chunk_iter = chunks.iter();
+        while let Some(chunk) = chunk_iter.next() {
+            if state == END && all_ascii_chunk(chunk) {
+                let skipped = skip_ascii_chunks(&mut chunk_iter);
+                pos += skipped * CHUNK_LEN;
+            } else {
+                state = dfa_run(state, chunk);
+                if unlikely(state == ERR) {
+                    break;
+                }
+                last_state = state;
+            }
+            pos += core::mem::size_of_val(chunk);
+        }
+        if unlikely(state == ERR) {
+            return Err(utf8_find_error(inp, pos, last_state != END));
+        }
+    }
+    // Did we leave the optimized loop in the middle of a UTF-8 sequence?
+    let was_mid_char = state != END;
+    debug_assert!(state != ERR);
+    if !tail.is_empty() {
+        // Check and early return if the last CHUNK_LEN bytes were all ASCII. The
```

Comment:
(note to self: reword this comment, since it seems like I got distracted halfway through it)